[JUJU-332] Fix programmatic errors in production

SimonRichardson opened the pull request:
Every so often we can end up with a nil message. Which either means we
dropped a message or we're attempting to process a message that isn't
there. This can lead to issues in people attempting to use the library.

To simplify the subscriber code for pubsub, I and @achilleasa
implemented a single linked list that we can enqueue and dequeue.

Removing the complexity of knowing when to request the next message. It
removes the ambiguity of knowing if you have a next message or not. The
trick of knowing when next is blocking or not is also gone, as we can
use the existing data channel. If that's already full (it's a buffered
channel of 1), then just offer a default case in the select to allow the
bypassing of the blocking operation.

Message pointers in the implementation is gone, you either have a
message or you don't. We no longer need to check for empty. And in
doing so, the implementation iterates all the pending handlers at onces,
without checking for data or next each iteration.

The benchmark performance is on par against the existing implementation.
Splitting hairs between implementations.

New:

     PASS: benchmarks_test.go:19: BenchmarkSuite.BenchmarkStructuredNoConversions      500000   4604 ns/op
     PASS: benchmarks_test.go:43: BenchmarkSuite.BenchmarkStructuredSerialize          100000   18301 ns/op

Old:

     PASS: benchmarks_test.go:19: BenchmarkSuite.BenchmarkStructuredNoConversions      500000   4781 ns/op
     PASS: benchmarks_test.go:43: BenchmarkSuite.BenchmarkStructuredSerialize          100000   18633 ns/op

SimonRichardson committed:
Fix programmatic errors in production

Every so often we can end up with a nil message. Which either means we
dropped a message or we're attempting to process a message that isn't
there. This can lead to issues in people attempting to use the library.

To simplify the subscriber code for pubsub, I and @achilleasa
implemented a single linked list that we can enqueue and dequeue.

Removing the complexity of knowing when to request the next message. It
removes the ambiguity of knowing if you have a next message or not. The
trick of knowing when next is blocking or not is also gone, as we can
use the existing data channel. If that's already full (it's a buffered
channel of 1), then just offer a default case in the select to allow the
bypassing of the blocking operation.

Message pointers in the implementation is gone, you either have a
message or you don't. We no longer need to check for empty. And in
doing so, the implementation iterates all the pending handlers at onces,
without checking for data or next each iteration.

The benchmark performance is on par against the existing implementation.
Splitting hairs between implementations.

New:

   PASS: benchmarks_test.go:19:
   BenchmarkSuite.BenchmarkStructuredNoConversions      500000
   4604 ns/op
   PASS: benchmarks_test.go:43:
   BenchmarkSuite.BenchmarkStructuredSerialize          100000
   18301 ns/op

Old:

   PASS: benchmarks_test.go:19:
   BenchmarkSuite.BenchmarkStructuredNoConversions      500000
   4781 ns/op
   PASS: benchmarks_test.go:43:
   BenchmarkSuite.BenchmarkStructuredSerialize          100000
   18633 ns/op

howbazaar commented on simplehub.go:
```diff
@@ -85,10 +85,10 @@ func (h *SimpleHub) Publish(topic string, data interface{}) func() {
+			s.notify(handlerCallback{
+				topic:  topic,
+				data:   data,
+				doneFn: wait.Done,
```
Nice change.

howbazaar reviewed (approved):
Seems reasonable to me.

SimonRichardson commented:
Hi @howbazaar, hope all is well 👌 thanks for reviewing this 👍

achilleasa reviewed (commented):
Changes LGTM

SimonRichardson commented:
$$merge$$

SimonRichardson commented:
!!build!!

SimonRichardson committed:
Fix go vet tooling

The tooling for go vet was old, updated to correctly work with the later
tool chain.

SimonRichardson commented:
$$merge$$
